src/Rings/FreeAssociativeAlgebraIdeal.jl

```diff
@@ -184,37 +193,65 @@ Ideal generating system with elements
 function groebner_basis(I::FreeAssociativeAlgebraIdeal,
   deg_bound::Int=-1;
   protocol::Bool=false,
-  interreduce::Bool=false)
+  interreduce::Bool=false,
+  algorithm::Symbol=:f4
+  )
   isdefined(I, :gb) && (I.deg_bound == -1 || I.deg_bound >= deg_bound) && return I.gb
-  I.gb = groebner_basis(IdealGens(gens(I)), deg_bound; ordering=:deglex, protocol=protocol, interreduce=interreduce)
+  I.gb = groebner_basis(IdealGens(gens(I)), deg_bound; ordering=:deglex, protocol=protocol, interreduce=interreduce, algorithm=algorithm)
   I.deg_bound = deg_bound
   return I.gb
 end
 function groebner_basis(g::IdealGens{<:FreeAssociativeAlgebraElem},
   deg_bound::Int=-1;
   ordering::Symbol=:deglex,
   protocol::Bool=false,
-  interreduce::Bool=false)
-  gb = groebner_basis(collect(g), deg_bound; ordering=ordering, protocol=protocol, interreduce=interreduce)
+  interreduce::Bool=false,
+  algorithm::Symbol=:f4
+  )
+  gb = groebner_basis(collect(g), deg_bound; ordering=ordering, protocol=protocol, interreduce=interreduce,algorithm=algorithm)
+  gb = Generic.FreeAssociativeAlgebraElem{QQFieldElem}[x for x in gb]
   return IdealGens(gb)
 end
 function groebner_basis(g::Vector{<:FreeAssociativeAlgebraElem},
   deg_bound::Int=-1;
   ordering::Symbol=:deglex,
   protocol::Bool=false,
-  interreduce::Bool=false)
+  interreduce::Bool=false,
+  algorithm::Symbol=:f4)
 
+  @req length(g) > 0 "Vector must not be empty"
   R = parent(g[1])
   @req all(x -> parent(x) == R, g) "parent mismatch"
-  @req deg_bound >= 0 || !protocol "computing with a protocol requires a degree bound"
+  #@req deg_bound >= 0 || !protocol "computing with a protocol requires a degree bound"
   @req ordering == :deglex || deg_bound > 0 "only :deglex ordering is supported for no degree bound"
+  @req algorithm in (:f4, :buchberger, :letterplace) "Not a valid algorithm, use :f4ncgb, :buchberger or :letterplace"
+
+  if algorithm == :f4
+    @req ordering == :deglex "f4 only supports :deglex ordering"
+    @req R.base_ring == QQ "only rational coefficients are supported"
+    protocol && f4ncgb_set_msg_printing(protocol)
+
+    handle = f4ncgb_init()
+    f4ncgb_set_nvars(handle, UInt32(ngens(R)))
+    f4ncgb_set_blocks(handle, UInt32[ngens(R)])
+    f4ncgb_set_threads(handle, UInt32(1))
+    if (deg_bound > 0) 
+      f4ncgb_set_maxdeg(handle, UInt32(deg_bound))
+    end
+    f4ncgb_add.(Ref(handle), g)
+    userdata = f4ncgb_polys_helper(R)
+    f4ncgb_solve(handle, userdata)
+    f4ncgb_free(handle)
```

Comment:
What happens if this line is left out, is data leaked? If so: what happens if a user presses Ctrl-C in the middle of that computation -- is there a leak then?